crates/typstyle-core/src/pretty/code_chain.rs

```diff
@@ -21,25 +27,68 @@ impl<'a> PrettyPrinter<'a> {
 
     /// Convert the node as dot chain, if in code, or in markup with at least two FieldAccess and one FuncCall.
     pub(super) fn try_convert_dot_chain(&'a self, node: &'a SyntaxNode) -> Option<ArenaDoc<'a>> {
-        if self.current_mode().is_code() {
-            return Some(self.convert_dot_chain(node));
-        } else if self.current_mode().is_markup() {
-            let mut chain_len = 0;
-            let mut has_call = false;
-            for node in resolve_dot_chain(node) {
-                if node.kind() == SyntaxKind::FieldAccess {
-                    chain_len += 1;
-                } else if node.kind() == SyntaxKind::FuncCall {
-                    has_call = true;
-                }
+        let mut dot_num = 0;
+        let mut call_num = 0;
+        let mut has_comment = false;
+        let chain: Vec<&SyntaxNode> = resolve_dot_chain(node).collect_vec();
+        for node in &chain {
+            if node.kind() == SyntaxKind::FieldAccess {
+                dot_num += 1;
+            } else if node.kind() == SyntaxKind::FuncCall {
+                call_num += 1;
             }
-            if chain_len > 1 && has_call {
-                return Some(self.parenthesize_if_necessary(|| self.convert_dot_chain(node)));
+            if has_comment_children(node) {
+                has_comment = true;
             }
         }
+        if dot_num > 1 && call_num == 1 && !has_comment {
+            if let Some(res) = self.try_convert_dot_chain_plain(chain) {
+                return Some(res);
+            }
+        }
+        if self.current_mode().is_markup() && dot_num > 1 && call_num > 0 {
+            return Some(self.parenthesize_if_necessary(|| self.convert_dot_chain(node)));
+        } else if self.current_mode().is_code() {
+            return Some(self.convert_dot_chain(node));
+        }
         None
     }
 
+    /// Used for ident.ident(args)
+    fn try_convert_dot_chain_plain(
+        &'a self,
+        mut chain: Vec<&'a SyntaxNode>,
+    ) -> Option<ArenaDoc<'a>> {
+        chain.reverse();
+        let func_call = chain.last()?.cast::<FuncCall>()?;
+        let ident = chain[0].cast::<Ident>()?;
+
+        // If the chain idents are long enough, do not put them in one line.
+        let estimated_len = ident.get().len()
+            + chain
+                .iter()
+                .skip(1)
```

Comment:
quick question: which item are we skipping?

Reply:
the starting ident